File: BaekJoon/백트래킹/15649_N과 M (1).cpp
#include <iostream>
#define MAX 9
using namespace std;

int n, m;
int visit[MAX] = { 0, };
int input[MAX] = { 0, };

void dfs(int cnt) {
	if (cnt == m) {
		for (int i = 0; i < m; i++) {
			cout << input[i] << " ";
		}
		cout << "\n";
		return;
	}
	for (int i = 1; i <= n; i++) {
		if (!visit[i]) { // 방문하지 않은 경우
			visit[i] = true;
			input[cnt] = i;
			dfs(cnt + 1);
			visit[i] = false;
		}
	}
}

int main() {
	cin >> n >> m;
	dfs(0);
}

File: BaekJoon/백트래킹/15649_N과 M (1).kt
import java.util.*

lateinit var visit: Array<Int>
lateinit var input: Array<Int>
var n: Int = 0
var m: Int = 0

fun main() = with(Scanner(System.`in`)) {
    n = nextInt()
    m = nextInt()

    visit = Array(n + 1) { 0 }
    input = Array(n + 1) { 0 }

    dfs(0)
}

fun dfs(depth: Int) {
    if (depth == m) {
        for (i in 0 until m) {
            print("${input[i]} ")
        }
        println()
        return
    }

    for (i in 0 until n) {
        if (visit[i] == 0) {
            visit[i] = 1

            input[depth] = i + 1
            dfs(depth + 1)

            visit[i] = 0
        }
    }
    return
}
